### ISSUE TITLE
Masking Layer results in additivity check failing

### ISSUE THREAD

**Issue by Zahlii** [shap/shap#1860]
```python
import tensorflow as tf
tf.__version__
```

    '2.4.1'

```python
import shap
shap.__version__
```

    '0.39.0'

```python
import numpy as np

C = 99

x = np.random.random((20, 30, 3))
y = x.sum(axis=-1, keepdims=True)
print(x.shape, y.shape)

# first 4 samples are truncated after step 25
x[0:4, 25:, :] = C

with tf.device("CPU:0"):
    class MaskedConv1D(tf.keras.layers.Conv1D):
        def compute_mask(self, input, mask=None):
            return mask

    i = tf.keras.layers.Input(shape=(30, 3))
    # m = tf.keras.layers.Masking(C)(i) # if you enable this line and remove the next, additivity fails
    m = i
    o = MaskedConv1D(use_bias=False, kernel_size=1, filters=1, kernel_initializer="ones")(m)
    
    model = tf.keras.models.Model(i, o)
    model.summary()
    model.compile(loss="MSE", optimizer="Adam")
    model.evaluate(x, y)
    
    fixed_out = tf.reduce_sum(o, axis=-1)
    s_model = tf.keras.models.Model(i, fixed_out)
    
    exp = shap.DeepExplainer(s_model, x)
    s = exp.shap_values(x) # this fails

    exp = shap.DeepExplainer(s_model, x[4:]) # excluding them from data AND from the explanations will work
    s = exp.shap_values(x[4:])
```

    (20, 30, 3) (20, 30, 1)
    Model: "model_11"
    _________________________________________________________________
    Layer (type)                 Output Shape              Param #   
    =================================================================
    input_9 (InputLayer)         [(None, 30, 3)]           0         
    _________________________________________________________________
    masked_conv1d_7 (MaskedConv1 (None, 30, 1)             3         
    =================================================================
    Total params: 3
    Trainable params: 3
    Non-trainable params: 0
    _________________________________________________________________
    1/1 [==============================] - 0s 84ms/step - loss: 2910.9321

Note that we can circumvent the problem by building the same model WITHOUT the mask layer, and then simply adding our custom masking and loading the weights from the existing model.

**Comment by Zahlii** [comment]
Possibly related: https://github.com/slundberg/shap/issues/213

**Comment by Zahlii** [comment]
As a temporary workaround, I wrote a function that will strip masking layers from my models, optionally handling the case that we have a time-based model with variable sizes (as this was essentially what I'm trying to do).

Maybe these could be an optional input to the DeepExplainer's TF implementation? @slundberg if you could give your thought on this, I can open up a PR on this topic.

- Stripping away masking layer
- Converting a (time-) dynamic model to a time-fixed model

```python
# Example:

C = 99

x = np.random.random((10, 30, 3))
y = x.sum(axis=-1, keepdims=True)

# first 4 samples are truncated after step 25
x[0:4, 25:, :] = C


class MaskedConv1D(tf.keras.layers.Conv1D):
    def compute_mask(self, input, mask=None):
        return mask

i = tf.keras.layers.Input(shape=(None, 3))
m = tf.keras.layers.Masking(C)(i)
o = MaskedConv1D(use_bias=False, kernel_size=1, filters=1, kernel_initializer="ones")(m)
o = tf.reduce_sum(o, axis=-1)
model = tf.keras.models.Model(i, o)

model.summary()
model.compile(loss="MSE", optimizer="Adam")

model.evaluate(x, y)

alt_model = set_fixed_time_dim(i, o, sequence_length=30, strip_masking=True, custom_objects={"MaskedConv1D": MaskedConv1D})
alt_model.summary()

p1 = alt_model.predict(x)
p2 = model.predict(x)

delta = p1 - p2
delta[0:4, 25:] = 0 # these will differ of course; but we can manually ignore them later
assert np.isclose(delta.sum(), 0)

exp = shap.DeepExplainer(alt_model, x)
s = exp.shap_values(x)


def set_fixed_time_dim(
    inputs: List, output, sequence_length: int, custom_objects=None, strip_masking=True
):
    """
    This is a very hacky way to convert a dynamic model with dynamic model input (time steps) into a fixed
    representation for e.g. SHAP values. This is done by creating a new model with changed input specs and then
    copying over weights. It also allows us to remove masking layers that currently do produce invalid SHAP values.
    See https://github.com/slundberg/shap/issues/1860

    Parameters
    ----------
    strip_masking
    inputs
    output
    sequence_length
    custom_objects

    Returns
    -------

    """
    old_model = tf.keras.models.Model(inputs, output)
    conf = old_model.get_config()
    replace_map = {}

    new_layers = []

    for i in conf["layers"]:
        if i["class_name"] == "InputLayer":
            s = list(i["config"]["batch_input_shape"])
            if len(s) == 3 and s[1] is None:
                s[1] = sequence_length
            i["config"]["batch_input_shape"] = tuple(s)
            new_layers.append(i)
        elif i["class_name"] == "Masking":
            replace_map[i["name"]] = i["inbound_nodes"][0][0][0]
            if not strip_masking:
                new_layers.append(i)
        else:
            if strip_masking and "inbound_nodes" in i and "TFOp" not in i["class_name"]:
                for n in i["inbound_nodes"][0]:
                    if n[0] in replace_map:
                        n[0] = replace_map[n[0]]
            new_layers.append(i)

    conf["layers"] = new_layers

    new_model = tf.keras.models.Model.from_config(conf, custom_objects=custom_objects)
    new_model.set_weights(old_model.get_weights())
    return new_model
```
